### ISSUE TITLE
Object references

### ISSUE THREAD

**Issue by soulnick** [gullerya/object-observer#15]
if you do this with standard object:
var t={a:{},b:{}}
undefined
t.a.c='i'
"i"
t.b=t.a
{c: "i"}
t.a.c
"i"
t.b.c
"i"
t.a.c='o'
"o"
t.a.c
"o"
t.b.c
"o"

with your object observer you will see that t.b.c. will be "i".
Because when you do t.b=t.a you clone t.a.
I think in this case we don't need to clone but to add another parent to the observer.
It will complicate a lot the code but it will be as close as possible to a standard js object
Thank you.
ps I think your work is huge
ps2 revoke oldvalue after deliver the changes because if newvalue is a child of oldvalue it will throw an error

**Comment by gullerya** [comment]
Hi,

First - thanks for the kudos down there.

Regarding the object referentiality - I see your point, you are right. And indeed it is either to preserve standard objects behavior (should still see what would be the price) or allow some limitation here. I prefer, of course, to go first - let me think about it.

Regarding the revokation - looks also right to me, I'll just verify that i'm not missing something that was intentionally making me do it at earliest point and if none - will fix it.

**Comment by gullerya** [comment]
Hi,

I've played around the issues that you've raised here and there, it took me some time, but finally I have something to say about them (hopefully something smart :)).

Let's start from the last one, the remark that if a child object is replacing the parent in the graph.
I've opened another defect with the precise definition of the issue [here](https://github.com/gullerya/object-observer/issues/18). Thanks for this one!

Regarding the preservation of referential correctness things are getting complicate. :)
It is indeed quite a work to support it and most concerning part here is that it will certainly affect the performance of the processing of any case, even when object has a single parent, which is a vast majority of cases.
Beside that I've realized, that even if I can practically provide a referential correctness for the use-case you gave - same object appearing N times in the single ancestor's graph, it is almost impossible to provide the same functionality across the graphs. At least impossible without some serious memory concerns and/or touching the original object that is an input for an observation. I'd like to refrain from both of those.

From the other side, I think that in the contexts where `object-observer` is applicable, having same objects duplicated more than once on the graph is very uncommon.

Bottom line: I'd like to postpone handling such a case and meanwhile only make it explicitly visible that cloning creates object duplication when same reference found more than once on the graph.
I'm currently trying to decide to throw an error when encountering such a case or write some warning to the log.

WDYT of all of this?

**Comment by gullerya** [comment]
Closing this issue due to quite a big change required while not yet proven strong requirement from the real-life usage, please re-open if any new considerations arose.

**Comment by gullerya** [comment]
I'm reopening this issue for a tracking and will attempt to look onto it again

**Comment by gullerya** [comment]
After a 'fresh' look onto the issue i'm again postponing to deal with it - the caveats are managing multiple parent contexts per observable which will definitely have a negative impact on performance and code complexity/maintanability.
Let me wait for a real need for this one, if any (there were no public signs of interest except this thread, purely tech one).
